torchao/dtypes/uintx/gemlite_layout.py

```diff
@@ -0,0 +1,372 @@
+                assert step == 1, "Only step == 1 is supported in slicing right now"
+                data_len = int_data.shape[dim]
+                # scale and zero_point are transposed compared to int_data
+                param_dim = 1 - dim
+                scale_len = scale.shape[param_dim]
+                ratio = data_len / scale_len
+                start_scale = int(start / ratio)
+                end_scale = int(end / ratio)
+
+                int_data = aten.slice.Tensor(int_data, dim, start, end, step)
+                # this is to handle padding
+                scale = aten.slice.Tensor(
+                    scale, param_dim, start_scale, end_scale, step
+                )
+                if zero_point is not None and zero_point.numel() > 0:
+                    zero_point = aten.slice.Tensor(
+                        zero_point, param_dim, start_scale, end_scale, step
+                    )
+                else:
+                    zero_point = None
+                # import fbvscode; fbvscode.set_trace()
+                sliced = self.from_plain(int_data, scale, zero_point, self._layout)
+                return sliced
+            else:
+                raise NotImplementedError(
+                    f"GemliteAQTTensorImpl dispatch: attempting to run {func}, with dim={dim}, that is not supported"
+                )
+
+        raise NotImplementedError(
+            f"GemliteAQTTensorImpl dispatch: attempting to run {func}, this is not supported"
+        )
+
+    __torch_function__ = torch._C._disabled_torch_function_impl
+
+    def get_layout(self) -> Layout:
+        return self._layout
+
+
+# logic taken from gemlite's core.py
+def _matmul_type_fn(batch_size: int, bit_width: int) -> str:
+    if batch_size > 64:
+        return "GEMM"
+    elif batch_size > 1:
+        return "GEMM_SPLITK"
+    elif bit_width < 8:
+        return "GEMV_REVSPLITK"
+    else:
+        return "GEMV_SPLITK"
+
+
+def _linear_fp_act_int4_weight_gemlite_impl(input_tensor, weight_tensor, bias=None):
+    if hasattr(weight_tensor, "tensor_impl"):
+        weight_impl = weight_tensor.tensor_impl
+    else:
+        weight_impl = weight_tensor
+
+    from gemlite.core import GemLiteLinearTriton
+
+    batch_size = input_tensor.view(-1, input_tensor.shape[-1]).shape[0]
+    matmul_type = _matmul_type_fn(batch_size, weight_impl._layout.bit_width)
+
+    return GemLiteLinearTriton.forward_functional(
+        x=input_tensor,
+        bias=bias,
+        matmul_type=matmul_type,
+        **weight_impl.gemlite_kwargs,
+        tensor_args=(
+            weight_impl.packed_weight,
+            weight_impl.scale,
+            weight_impl.zero_point,
+        ),
+    )
+
+
+def _linear_fp_act_int4_weight_gemlite_check(input_tensor, weight_tensor, bias):
+    return (
+        # input is native fp16 tensor
+        not is_traceable_wrapper_subclass(input_tensor)
+        # and input_tensor.dtype == torch.float16
```

Comment:
don't we need this?

Reply:
its a little redundant since we block creation of the subclass unpon creation, normally when you mess this type of thing up it gives you a dtype mismatch error which is what will happen now, whereas adding this would likely bypass the error which i don't know if users would actually want.

could see it being either way though.